[groupepsa] Initial Contribution

arjanmels opened the pull request:
New Binding for Groupe PSA (Opel, Peugeot, Citroen, DS, Vauxhall) cars

Binding to retrieve information via the Groupe PSA Web API for cars.

A compiled version is available at: https://github.com/arjanmels/openhab2-addons/releases

Forum thread: https://community.openhab.org/t/new-binding-for-groupe-psa-peugeot-citroen-ds-opel-vauxhall-cars/110580/7

Testing by various forum members has taken place. 

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It is a good practice to add a URL to your built JAR in this Pull Request description,
so it is easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

Skinah commented:
FYI the channels need to be lowerCamelCase, see here:
https://www.openhab.org/docs/developer/guidelines.html#java-coding-style

arjanmels commented:
@Skinah Thank you. Corrected in latest update.

SamuelBrucksch commented:
Can you please supply a new compiled version of this so i can test it? The last version did not work for me.

SamuelBrucksch commented:
This contains 3309 changed files, i did not read all of the communication in here, if this was adressed already please ignore my comment, but if not... maybe this should be cleaned up and be reopened? Will be quite hard to review i think.

Skinah commented:
You should be able to fix it by...

```
git rebase -i HEAD~1
```

replace the 1 with how many commits you wish to roll back on. Then you can press the D button to drop/delete the commits that are not yours. After doing this you do a force push. Its a pain if it is hundreds of commits but its just pressing D to drop them so it can be fixed in 10 minutes this way.

arjanmels commented:
@Skinah Thanks for the advise. Finally found time to work on this again. Rebased successfully and also corrected commit signatures. Please let me know if other changes need to be done (not certain why the CI / Build fails).

wborn commented:
> not certain why the CI / Build fails

Probably because your add-on has not yet been added to `bom/openhab-addons/pom.xml`.

arjanmels commented:
@wborn @Skinah Thanks for the advise. I added the binding to the bom pom.xml and cleaned up codestyle review errors and warnings (copyright was -2020 instead of -2021).

Ready for further review I hope :-)

fwolter commented:
@arjanmels What's the state of this PR? There are so many events in this PR, that GitHub hides the most relevant ones.

arjanmels commented:
@fwolter this one is good to go from my perspective. (I properly rebased it last month.)

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/GroupePSAHandlerFactory.java:
```diff
@@ -0,0 +1,99 @@
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_BRIDGE, THING_TYPE_VEHICLE);
+
+    private final OAuthFactory oAuthFactory;
+    protected final @NonNullByDefault({}) HttpClient httpClient;
```
The null annotation can be removed as you inject the reference via the constructor.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/GroupePSAHandlerFactory.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    @Override
+    protected synchronized void removeHandler(ThingHandler thingHandler) {
```
Did you take a look at the `ThingHandlerService`? That would make the code much simpler. See https://www.openhab.org/docs/developer/bindings/#discovery-that-is-bound-to-a-bridge

arjanmels replied:
Done

fwolter commented on bundles/org.openhab.binding.groupepsa/README.md:
```diff
@@ -0,0 +1,28 @@
+
+Use the "Groupe PSA Web Api bridge" to auto discover your cars. You need to select the brand for the bridge binding and only cars for the brand will be auto discovered. If you need to add for multiple brands or multiple different users, add multiple bridges.
+
+## Binding Configuration
```
If I see correctly, you binding doesn't have a configuration at binding level. You might mean the bridge configuration.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/bridge/GroupePSABridgeHandler.java:
```diff
@@ -0,0 +1,214 @@
+
+    private final OAuthFactory oAuthFactory;
+
+    private @NonNullByDefault({}) OAuthClientService oAuthService;
```
This cannot be null, so the null annotation can be removed.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/bridge/GroupePSABridgeHandler.java:
```diff
@@ -0,0 +1,214 @@
+    private @Nullable ScheduledFuture<?> groupepsaBridgePollingJob;
+    private final HttpClient httpClient;
+
+    private @NonNullByDefault({}) String vendor;
```
You could initialize the attributes in the configuration with a default value and remove their nullable annotation. Then, you could remove this annotation.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/bridge/GroupePSABridgeHandler.java:
```diff
@@ -0,0 +1,214 @@
+    private @NonNullByDefault({}) String clientId;
+    private @NonNullByDefault({}) String clientSecret;
+
+    private @NonNullByDefault({}) GroupePSAConnectApi groupePSAApi;
```
This should be `@Nullable`. Disabling the null annotations with `@NonNullByDefault({}` should be used only if you have good reasons. Please check all.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/bridge/GroupePSABridgeHandler.java:
```diff
@@ -0,0 +1,214 @@
+            List<Vehicle> vehicles = getVehicles();
+            if (vehicles != null) {
+                updateStatus(ThingStatus.ONLINE);
+                logger.debug("Found {} vehicles", vehicles.size());
+                for (Vehicle vehicle : vehicles) {
+                    logger.trace("Vehicle: {}", vehicle);
+                }
```
Can this message be replaced by using the debugger? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/bridge/GroupePSABridgeHandler.java:
```diff
@@ -0,0 +1,214 @@
+                    logger.trace("Vehicle: {}", vehicle);
+                }
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "@text/comm-error-query-vehicles-failed");
+                logger.warn("Unable to fetch vehicles");
```
The log message could be removed, as the status update is already logged by the framework.

The state change originated by updateStatus() is logged to events.log. Including the status detail message.

Please check all.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/discovery/GroupePSADiscoveryService.java:
```diff
@@ -0,0 +1,94 @@
+                }
+            }
+        } catch (GroupePSACommunicationException e) {
+            logger.warn("No vehicles found", e);
```
Are you logging the stack trace by intention?

arjanmels replied:
Yes, there are various reasons why the GroupePSACommunicationException can be thrown.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/rest/api/GroupePSAConnectApi.java:
```diff
@@ -0,0 +1,200 @@
+
+        try {
+            ContentResponse response = request.send();
+            logger.trace("HttpResponse {}", response);
+            logger.trace("HttpResponse Headers:\n{}", response.getHeaders());
+            logger.trace("HttpResponse Content: {}", response.getContentAsString());
```
Can this message be replaced by using the debugger? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

arjanmels replied:
Prefer to keep it in as it helps debugging other peoples setups.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/things/GroupePSAConfiguration.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * The {@link GroupePSAConfiguration} class contains fields mapping thing configuration parameters.
+ *
+ * @author Markus Pfleger - Initial contribution
```
Is the author correct?

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/things/GroupePSAHandler.java:
```diff
@@ -0,0 +1,487 @@
+        }
+    };
+
+    public GroupePSAHandler(Thing thing) {
+        super(thing);
+    }
```
What's the purpose of this?

arjanmels replied:
Otherwise I get: Implicit super constructor org.openhab.core.thing.binding.BaseThingHandler() is undefined for default constructor. Must define an explicit constructor

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/things/GroupePSAHandler.java:
```diff
@@ -0,0 +1,487 @@
+    }
+
+    private synchronized void updateGroupePSAState() {
+        if (System.nanoTime() - lastQueryTimeMs <= maxQueryFrequencyNanos) {
```
Did you take a look at `ExpiringCache`? https://next.openhab.org/javadoc/latest/org/openhab/core/cache/expiringcache

arjanmels replied:
Thanks for the pointer. I think it would become a bit artificial to use it: I don't need to cache the actual value, but I want to protect it being retrieved too often (more then once a minute).

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/java/org/openhab/binding/groupepsa/internal/things/GroupePSAHandler.java:
```diff
@@ -0,0 +1,487 @@
+                        "@text/comm-error-query-vehicle-failed");
+            }
+        } catch (GroupePSACommunicationException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "@text/comm-error-query-vehicle-failed");
+            logger.warn("Unable to query groupepsa status for:  {}. Error: {}", id, e.getMessage());
```
The log message could be removed, as the status update is already logged by the framework.

You could include the exception's message in the thing status detail message to display it in the UI.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+
+	<name>Groupe PSA Binding</name>
+	<description>Provides information and (limited) control of cars from the Groupe PSA (Peugeot, Citroen, Opel, etc.).</description>
+	<author>Arjan Mels</author>
```
The author tag is deprecated and should therefore be removed. See openhab/openhab-core#1844.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/i18n/groupepsa.properties:
```diff
@@ -0,0 +1,32 @@
+binding.groupepsa.name = Groupe PSA Binding
+binding.groupepsa.description = Provides information and (limited) control of cars from the Groupe PSA (Peugeot, Citroen, Opel, etc.).
+
+# thing types
+#thing-type.groupepsa.sample.label = <Your localized Thing label>
+#thing-type.groupepsa.sample.description = <Your localized Thing description>
+
+# thing type config description
+#thing-type.config.groupepsa.sample.config1.label = <Your localized config parameter label>
+#thing-type.config.groupepsa.sample.config1.description = <Your localized config parameter description>
+
+# channel types
+#channel-type.groupepsa.sample-channel.label = <Your localized Channel label>
+#channel-type.groupepsa.sample-channel.description = <Your localized Channel description>
```
Can this be removed?

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,559 @@
+
+
+	<bridge-type id="bridge">
+		<label>Groupe PSA Web Api Bridge</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,559 @@
+	<channel-group-type id="environment">
+		<label>Environment</label>
+		<description>Environmental conditions</description>
+		<category>Sun_Clouds</category>
```
Are you sure this category exists? It's not listed here https://www.openhab.org/docs/concepts/categories.html#thing-categories

arjanmels replied:
It is an existing channel category, I think: https://www.openhab.org/docs/concepts/categories.html#weather

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,559 @@
+		<item-type>Contact</item-type>
+		<label>Daytime</label>
+		<description>Enabled if it is daytime</description>
+		<category>Sun</category>
```
See above. Not listed https://www.openhab.org/docs/concepts/categories.html#thing-categories

arjanmels replied:
It is an existing channel category, I think: https://www.openhab.org/docs/concepts/categories.html#weather

fwolter replied:
Sure!

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,559 @@
+
+	<channel-type id="doorLock">
+		<item-type>String</item-type>
+		<label>Door lock</label>
```
Words in labels should be capitalized (except prepositions and so on). See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,559 @@
+
+	<channel-type id="doorOpen">
+		<item-type>Contact</item-type>
+		<label>Door open</label>
```
Please check all.
```suggestion
		<label>Door Open</label>
```

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,559 @@
+		<item-type>String</item-type>
+		<label>Ignition</label>
+		<description>Ignition state</description>
+		<category>Motion</category>
```
See above. Not listed. Please check all.

arjanmels replied:
It is a channel category, I believe: https://www.openhab.org/docs/concepts/categories.html#properties

fwolter replied:
You're right, but I don't see the relation to the Motion category. This category is more intended for motion or presence detectors.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,559 @@
+		</state>
+	</channel-type>
+
+	<channel-type id="signal">
+		<item-type>Number:Dimensionless</item-type>
```
There's a predefined signal strength channel you could use. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#system-state-channel-types

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,559 @@
+		</state>
+	</channel-type>
+
+	<channel-type id="service">
+		<item-type>String</item-type>
+		<label>Service</label>
+		<description>Service Type</description>
+		<state readOnly="true">
+			<options>
+				<option value="Electric">Electric</option>
+				<option value="Hybrid">Hybrid</option>
+				<option value="Unknown">Unknown</option>
```
This seems rather a property than a Channel.

arjanmels replied:
I'm not entirely certain. It may indeed be, but it could be that it theoretically changes if the car is driving on electricity or on fuel. I propose to keep it a channel just in case.
(There are quite some channels that don't work on all cars, so I can't test it properly.)

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,559 @@
+		<label>Charging Time Remaining</label>
+		<description>Time remaining till charged</description>
+		<category>Time</category>
+		<state pattern="%tT"/>
```
Are you sure this formats nicely?

arjanmels replied:
It gives me full date and time. For my Dutch system: "zondag, 24 apr. 2022 18:00".

fwolter reviewed (changes requested):
Thanks for your contribution! I reviewed your code and here is my feedback.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html` after building it with `mvn clean install`.

Please add yourself to the CODEOWNERS file.

arjanmels commented:
@fwolter Thank you for your very thorough review. I'll try to get round to the changes in the next 2 weeks.

arjanmels commented:
@fwolter Finally got round to this binding again. I addressed your suggestions. Looking forward to further review.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,551 @@
+		<item-type>Contact</item-type>
+		<label>Moving</label>
+		<description>Vehicle is moving</description>
+		<category>Motion</category>
```
You don't need to put a category forcibly to any Channel. If there's no suitable category, better leave it away.

arjanmels replied:
Ok, will only leave clearly relevant ones in.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,551 @@
+	<channel-type id="preconditioningFailure">
+		<item-type>String</item-type>
+		<label>Air Conditioning Failure</label>
+		<description>Airr conditioning failure cause</description>
```
```suggestion
		<description>Air conditioning failure cause</description>
```

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,551 @@
+	<channel-group-type id="electric">
+		<label>Electric Energy</label>
+		<description>Electric energy</description>
+		<category>Electric motor energy status</category>
```
This doesn't seem a valid category.

fwolter commented on bundles/org.openhab.binding.groupepsa/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,551 @@
+		<item-type>Number:Dimensionless</item-type>
+		<label>Fuel Level</label>
+		<description>Fuel level</description>
+		<category>Pressure</category>
```
What's the relation to pressure? Same for below.

arjanmels replied:
I just liked the icon. Will remove.

fwolter reviewed (changes requested):
There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html` after building it with `mvn clean install`.

arjanmels commented on bundles/org.openhab.binding.groupepsa/README.md:
```diff
@@ -0,0 +1,151 @@
+chargingStatus|String|Battery charging status
+chargingMode|String|Battery charging mode
+chargingPlugged|Contact|Vehicle plugged in to charger
+chargingRate|Number:Speed|Battery Charging Rate
```
The charging rate is expressed as km/h (added range per hour).

arjanmels commented on bundles/org.openhab.binding.groupepsa/README.md:
````diff
@@ -0,0 +1,151 @@
+
+### Things file
+
+```perl
````
Ok, thanks. Highlighting in visual code was nicer with the language specifier.

arjanmels commented on bundles/org.openhab.binding.groupepsa/README.md:
```diff
@@ -0,0 +1,151 @@
+|userName|None|Yes|The user name for the mypeugot/mycitroen/myds/myopel/myvauxhall website or app|
+|password|None|Yes|The password for the given user|
+|pollingInterval|60|No|The Polling interval (in minutes) determines how often the available vehicles are queried|
+|clientId||Yes|The client ID for API access: can normally left at the default value. (see: https://developer.groupe-psa.io/webapi/b2c/quickstart/connect/#article)|
+|clientSecret||Yes|The client secret for API access: can normally left at the default value. (see: https://developer.groupe-psa.io/webapi/b2c/quickstart/connect/#article)|
```
These are the credentials also used by the mobile apps. (Creating individual API keys with the proper permissions is quite an involved process.)

fwolter reviewed (approved):
LGTM
Now, you could add your binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website

arjanmels commented:
Thank you! I just submitted a pull request for the logo.
